### PR TITLE
Demonstrating some stuff which does not work as I would expect

### PR DESCRIPTION
**Pull request by jglick**
Initial integration tests with https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/10 failed, but this seems to be due to preëxisting bugs here.

@reviewbybees

### CODE REVIEW

**Comment by abayer** [comment]
Thanks - I'll see what I can do.

File: src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java [hunk 1]
```diff
@@ -150,5 +150,13 @@ public void globalLibrarySuccess() throws Exception {
 
         j.assertLogContains("[nothing here]", b);
         j.assertLogContains("map call(1,2)", b);
+        /* TODO called the wrong way: closure1([a:1])
```
Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java]: So this is getting called [here](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/6f7cf8ed718fde0e8da3b31ed1235b15fbe6bd37/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy#L76-L79) - we're delegating the `Closure` to the `CpsScript` instance. I honestly don't know why that would be different in execution than normal, but it apparently is? Any ideas?

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java]: fwiw, I tried this as just a simple job with a function in the Pipeline -
```groovy
closure1(a: 1) { echo "Yo" }

def closure1(a, Closure body) { echo "closure1($a)"; body() }
```

...and it did the same thing.

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java]: And with 
```groovy
closure1(a: 1) { echo "Yo" }

closure2(a: 1, b: 2) {
    echo 'running inside closure2'
}

def closure1(a, Closure body) { echo "closure1($a)"; body() }

def closure2(a, b, Closure body) { echo "closure2($a, $b)"; body() }
```

I get:
```
Started by user admin
[Pipeline] echo
closure1([a:1])
[Pipeline] echo
Yo
[Pipeline] End of Pipeline
java.lang.NoSuchMethodError: No such DSL method 'closure2' found among steps [archive, bat, catchError, checkout, deleteDir, dir, dockerFingerprintFrom, dockerFingerprintRun, echo, envVarsForTool, error, fileExists, git, isUnix, load, mail, node, parallel, properties, pwd, readFile, readTrusted, retry, script, sh, sleep, stage, stash, step, timeout, tool, unarchive, unstash, waitUntil, withDockerContainer, withDockerRegistry, withDockerServer, withEnv, wrap, writeFile, ws] or symbols [all, always, apiToken, architecture, archiveArtifacts, artifactManager, batchFile, booleanParam, buildButton, buildDiscarder, caseInsensitive, caseSensitive, choice, choiceParam, clock, cloud, command, cron, crumb, defaultView, demand, downloadSettings, downstream, dumb, envVars, file, fileParam, filePath, fingerprint, frameOptions, freeStyle, freeStyleJob, hyperlink, hyperlinkToModels, installSource, jdk, jdkInstaller, jnlp, jobName, lastDuration, lastFailure, lastGrantedAuthorities, lastStable, lastSuccess, legacy, list, local, location, logRotator, loggedInUsersCanDoAnything, masterBuild, maven, maven3Mojos, mavenErrors, mavenMojos, mavenWarnings, myView, nodeProperties, nonStoredPasswordParam, none, paneStatus, parameters, password, pattern, plainText, plugin, projectNamingStrategy, proxy, queueItemAuthenticator, quietPeriod, run, runParam, schedule, scm, scmRetryCount, search, security, shell, slave, stackTrace, standard, status, string, stringParam, swapSpace, text, textParam, tmpSpace, toolLocation, unsecured, upstream, viewsTabBar, weather, zfs, zip]
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:145)
	at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:117)
	at groovy.lang.GroovyObject$invokeMethod.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:151)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:21)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:115)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:103)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:149)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:146)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:16)
	at WorkflowScript.run(WorkflowScript:3)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:48)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
	at sun.reflect.GeneratedMethodAccessor151.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ClosureBlock.eval(ClosureBlock.java:46)
	at com.cloudbees.groovy.cps.Next.step(Next.java:58)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:154)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:33)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:164)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:361)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:80)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:236)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:226)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:47)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Finished: FAILURE
```

I'm suspecting this is what's happening in the test here for closure2 but the `NoSuchMethodError` is getting swallowed...

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java]: And running the same code (i.e., still from raw Pipeline, not the model) but using it from the global library gets me 
```
hudson.remoting.ProxyException: groovy.lang.MissingMethodException: No signature of method: acmeFuncClosure2.call() is applicable for argument types: (java.util.LinkedHashMap, org.jenkinsci.plugins.workflow.cps.CpsClosure2) values: [[a:1, b:2], org.jenkinsci.plugins.workflow.cps.CpsClosure2@ed5c05c]
```
So...yeah. That's busted weird.

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java]: Running via a global library goes through a different code path. The symbol is looked up, then it is `call`ed. This is very different from having a definition inside the same script.

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java]: More from that stacktrace:

```
hudson.remoting.ProxyException: groovy.lang.MissingMethodException: No signature of method: acmeFuncClosure2.call() is applicable for argument types: (java.util.LinkedHashMap, org.jenkinsci.plugins.workflow.cps.CpsClosure2) values: [[a:1, b:2], org.jenkinsci.plugins.workflow.cps.CpsClosure2@7dd50d79]
Possible solutions: call(java.lang.Object, java.lang.Object, groovy.lang.Closure), wait(), any(), run(), run(), find()
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:58)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:64)
	at groovy.lang.GroovyObject$invokeMethod.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:151)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:21)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:115)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:103)
(etc)
```
So it's not ever even getting to `CpsScript.invokeMethod` when it's in the global library. WEIRD.

**Comment by abayer** [comment]
:bee: - merging this in for further work.

**Pull request merged**